onnx/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/onnx/OnnxInferenceModel.kt

```diff
@@ -87,8 +97,69 @@ public open class OnnxInferenceModel : InferenceModel() {
             }
             model.outputDataType = outputTensorInfo.type
 
+            model.pathToModel = pathToModel
+
             return model
         }
+
+    }
+
+    /**
+     * By default, the model is initialized with CPU execution provider with BFCArena memory allocator.
+     * This method allows to set the execution provider to use.
+     * If the model is already initialized, internal session will be closed and new one will be created.
+     * If [executionProvidersInUse] is the same as the one passed, nothing will happen.
+     * If execution provider is not supported, an exception will be thrown.
+     * @param executionProviders list of execution providers to use.
+     */
+    public fun reinitializeWith(vararg executionProviders: ExecutionProvider) {
+        require(::env.isInitialized) { "The model $this is not initialized!" }
+        require(executionProviders.isNotEmpty()) { "Please pass at least one execution provider!" }
+        for (executionProvider in executionProviders) {
+            require(executionProvider.internalProviderId in OrtEnvironment.getAvailableProviders()) {
+                "The optimized execution provider $executionProvider is not available in the current environment!"
+            }
+        }
+
+        val uniqueProviders = executionProviders.distinctBy { it.internalProviderId }
+
+        if (::executionProvidersInUse.isInitialized && uniqueProviders == executionProvidersInUse) {
+            return
+        }
+
+        if (::session.isInitialized) {
+            session.close()
+        }
+
+        val sessionOptions = SessionOptions()
+        for (provider in uniqueProviders) {
+            when (provider) {
+                is CUDA -> {
+                    sessionOptions.addCUDA(provider.deviceId)
+                }
+                is CPU -> {
+                     /*
+                         Doing nothing here because the CPU provider is always added to the sessionOptions
+                         after all other providers are added.
+                      */
+                }
+            }
+        }
+
+        /*
+            Users can explicitly add CPU provider with BFCArenaAllocator disabled,
+            but if not present, it will be added automatically with BFCArenaAllocator enabled.
+         */
+        uniqueProviders.find { it is CPU }?.let {
+            it as CPU
+            sessionOptions.addCPU(it.useBFCArenaAllocator)
+        } ?: run {
+            sessionOptions.addCPU(true)
+        }
+
+        session = env.createSession(pathToModel, sessionOptions)
```

Comment:
I'm not talking about changing the path, but about changes in the model file. There is no guarantee that this would not happen accidentally (and also someone could just try to misuse the api). I think that `initInputOutputInfo` should be called inside `reinitializeWith` function to ensure that input/output would be consistent with the model all the time.

Reply:
Interestingly, calling `initInputOutputInfo` twice would not change the shapes, because there are checks to initialize them twice. And we need the check for the input shape because for some models it is set manually before initialization. Perhaps, this is something that should be looked into later.